### PR TITLE
Improved socket implementation to recycle buffer in Socket.readXX ope…

### PR DESCRIPTION
**Pull request by pchlupacek**
…rations

### CODE REVIEW

**Comment by pchlupacek** [comment]
@mpilquist this complains about the binary compatibility, but hence the change is not in api, I still believe its ok to publish it in 0.9.3 any thoughts?

**Comment by mpilquist** [comment]
@pchlupacek Yeah that's fine -- can you add an exclusion to `mimaBinaryIssueFilters`?

File: io/src/main/scala/fs2/io/tcp/Socket.scala [hunk 1]
```diff
@@ -177,90 +177,128 @@ protected[tcp] object Socket {
...
-      def go(buff: ByteBuffer, timeoutMs:Long):F[Option[Chunk[Byte]]] = {
-        readChunk(buff,timeoutMs).flatMap { case (readBytes, took) =>
-          if (readBytes < 0 || buff.remaining() <= 0) F.pure(Some(Chunk.bytes(buff.array(), 0, buff.position())))
-          else go(buff,(timeoutMs - took) max 0)
+      def read0(max:Int, timeout:Option[FiniteDuration]):F[Option[Chunk[Byte]]] = {
+        readSemaphore.decrement >>
+        getBufferOf(max) flatMap { buff =>
+          readChunk(buff, timeout.map(_.toMillis).getOrElse(0l)) flatMap {
+            case (read, _) =>
+              if (read < 0) F.pure(None)
```
Comment (mpilquist) [inline io/src/main/scala/fs2/io/tcp/Socket.scala]: The `readSemaphore` isn't incremented here -- is that okay?

File: io/src/main/scala/fs2/io/tcp/Socket.scala [hunk 2]
```diff
@@ -177,90 +177,128 @@ protected[tcp] object Socket {
...
       }
-    }
 
-    def readN0(max:Int, timeout:Option[FiniteDuration]):F[Option[Chunk[Byte]]] = {
-      def go(buff: ByteBuffer, timeoutMs:Long):F[Option[Chunk[Byte]]] = {
-        readChunk(buff,timeoutMs).flatMap { case (readBytes, took) =>
-          if (readBytes < 0 || buff.remaining() <= 0) F.pure(Some(Chunk.bytes(buff.array(), 0, buff.position())))
-          else go(buff,(timeoutMs - took) max 0)
+      def read0(max:Int, timeout:Option[FiniteDuration]):F[Option[Chunk[Byte]]] = {
+        readSemaphore.decrement >>
```
Comment (mpilquist) [inline io/src/main/scala/fs2/io/tcp/Socket.scala]: Do we need any error handling here? E.g., if one of these tasks throws, we still want to increment the semaphore, right?

File: io/src/main/scala/fs2/io/tcp/Socket.scala (hunk 1 above)
Comment (pchlupacek) [inline io/src/main/scala/fs2/io/tcp/Socket.scala]: thanks for spotting this

Comment (pchlupacek) [inline io/src/main/scala/fs2/io/tcp/Socket.scala]: but actually, hence this signals eof, i think we won't hit it,neverthelss I gonna fix it for corectness

File: io/src/main/scala/fs2/io/tcp/Socket.scala (hunk 2 above)
Comment (pchlupacek) [inline io/src/main/scala/fs2/io/tcp/Socket.scala]: Well I think if you fail the read, there is no way to recover it anyhow, so my thought was just to let it be. Do you see there is a way that read will fail, and there is chance to recover it with next read?

**Comment by pchlupacek** [comment]
@mpilquist made modifications as you have suggested. Not sure about this mimaBinaryIssueFilters where they are?

**Comment by mpilquist** [comment]
@pchlupacek I took care of the mima issue: https://github.com/functional-streams-for-scala/fs2/commit/b02fd92556ef66d7216ceb8cb21fc0ec9e9e7e62

**Comment by pchlupacek** [comment]
@mpilquist thanks

**Pull request merged**